program/rust/src/rust_oracle.rs

```diff
@@ -23,4 +51,105 @@ pub fn update_version(
+                && valid_fresh_account(y) =>
+        {
+            Ok([x, y])
+        }
+        _ => Err(ProgramError::InvalidArgument),
+    }?;
+
+    // Initialize by setting to zero again (just in case) and populating the account header
+    clear_account(fresh_mapping_account)?;
+
+    let hdr = load::<cmd_hdr_t>(instruction_data);
+    {
+        let mut mapping_account = load_account_as_mut::<pc_map_table_t>(fresh_mapping_account)?;
+        mapping_account.magic_ = PC_MAGIC;
+        mapping_account.ver_ = hdr.ver_;
+        mapping_account.type_ = PC_ACCTYPE_MAPPING;
+        mapping_account.size_ =
+            (size_of::<pc_map_table_t>() - size_of_val(&mapping_account.prod_)) as u32;
+    }
+
+    Ok(SUCCESS)
+}
+
+fn valid_funding_account(account: &AccountInfo) -> bool {
+    account.is_signer && account.is_writable
+}
+
+fn valid_signable_account(program_id: &Pubkey, account: &AccountInfo, minimum_size: usize) -> bool {
+    account.is_signer
+        && account.is_writable
+        && account.owner == program_id
+        && account.data_len() >= minimum_size
+        && Rent::default().is_exempt(account.lamports(), account.data_len())
+}
+
+/// Returns `true` if the `account` is fresh, i.e., its data can be overwritten.
+/// Use this check to prevent accidentally overwriting accounts whose data is already populated.
+fn valid_fresh_account(account: &AccountInfo) -> bool {
+    let pyth_acc = load_account_as::<pc_acc>(account);
+    match pyth_acc {
+        Ok(pyth_acc) => pyth_acc.magic_ == 0 && pyth_acc.ver_ == 0,
+        Err(_) => false,
+    }
+}
+
+/// Sets the data of account to all-zero
+fn clear_account(account: &AccountInfo) -> Result<(), ProgramError> {
+    let mut data = account
+        .try_borrow_mut_data()
+        .map_err(|_| ProgramError::InvalidArgument)?;
+    let length = data.len();
+    sol_memset(data.borrow_mut(), 0, length);
+    Ok(())
+}
+
+/// Interpret the bytes in `data` as a value of type `T`
+fn load<T: Pod>(data: &[u8]) -> &T {
+    from_bytes(&data[0..size_of::<T>()])
+}
+
+/// Interpret the bytes in `data` as a mutable value of type `T`
+#[allow(unused)]
+fn load_mut<T: Pod>(data: &mut [u8]) -> &mut T {
+    from_bytes_mut(&mut data[0..size_of::<T>()])
+}
+
+/// Get the data stored in `account` as a value of type `T`
+fn load_account_as<'a, T: Pod>(account: &'a AccountInfo) -> Result<Ref<'a, T>, ProgramError> {
+    let data = account.try_borrow_data()?;
+
+    Ok(Ref::map(data, |data| {
+        bytemuck::from_bytes(&data[0..size_of::<T>()])
+    }))
+}
+
+/// Mutably borrow the data in `account` as a value of type `T`.
+/// Any mutations to the returned value will be reflected in the account data.
+fn load_account_as_mut<'a, T: Pod>(
+    account: &'a AccountInfo,
+) -> Result<RefMut<'a, T>, ProgramError> {
+    let data = account.try_borrow_mut_data()?;
+
+    Ok(RefMut::map(data, |data| {
+        bytemuck::from_bytes_mut(&mut data[0..size_of::<T>()])
```

Comment:
`try_from_bytes_mut` instead for error propagation

Reply:
for this one, i couldn't figure out how to swap the Ref<Result<...>> to Result<Ref<..>>

Reply:
I took a quick look, the only thing I could find is this:
https://stackoverflow.com/questions/72730465/how-do-i-propagate-an-error-outside-of-a-closure
Though not sure if it is worth the hastle, I think Guillermo or Tom might be more helpful. Specially Guillermo since he had a strong opinion on unwrap

I will check again tomorrow though